opm/simulators/linalg/bda/rocsparseSolverBackend.cpp

```diff
@@ -0,0 +1,626 @@
+            ", time per iteration: " << res.elapsed / it << ", iterations: " << it;
+        OpmLog::info(out.str());
+    }
+    if (verbosity >= 3) {
+        std::ostringstream out;
+        out << "rocsparseSolver::prec_apply:  " << t_prec.elapsed() << " s\n";
+        out << "rocsparseSolver::spmv:        " << t_spmv.elapsed() << " s\n";
+        out << "rocsparseSolver::rest:        " << t_rest.elapsed() << " s\n";
+        out << "rocsparseSolver::total_solve: " << res.elapsed << " s\n";
+        OpmLog::info(out.str());
+    }
+}
+
+
+template <unsigned int block_size>
+void rocsparseSolverBackend<block_size>::initialize(std::shared_ptr<BlockedMatrix> matrix, std::shared_ptr<BlockedMatrix> jacMatrix) {
+    this->Nb = matrix->Nb;
+    this->N = Nb * block_size;
+    this->nnzb = matrix->nnzbs;
+    this->nnz = nnzb * block_size * block_size;
+    nnzbs_prec = nnzb;
+
+    if (jacMatrix) {
+        useJacMatrix = true;
+        nnzbs_prec = jacMatrix->nnzbs;
+    }
+
+    std::ostringstream out;
+    out << "Initializing GPU, matrix size: " << Nb << " blockrows, nnzb: " << nnzb << "\n";
+    if (useJacMatrix) {
+        out << "Blocks in ILU matrix: " << jacMatrix->nnzbs << "\n";
+    }
+    out << "Maxit: " << maxit << std::scientific << ", tolerance: " << tolerance << "\n";
+    out << "PlatformID: " << platformID << ", deviceID: " << deviceID << "\n";
+    OpmLog::info(out.str());
+    out.str("");
+    out.clear();
+
+    mat = matrix;
+    jacMat = jacMatrix;
+
+    HIP_CHECK(hipMalloc((void**)&d_r, sizeof(double) * N));
+    HIP_CHECK(hipMalloc((void**)&d_rw, sizeof(double) * N));
+    HIP_CHECK(hipMalloc((void**)&d_p, sizeof(double) * N));
+    HIP_CHECK(hipMalloc((void**)&d_pw, sizeof(double) * N));
+    HIP_CHECK(hipMalloc((void**)&d_s, sizeof(double) * N));
+    HIP_CHECK(hipMalloc((void**)&d_t, sizeof(double) * N));
+    HIP_CHECK(hipMalloc((void**)&d_v, sizeof(double) * N));
+
+    HIP_CHECK(hipMalloc((void**)&d_Arows, sizeof(rocsparse_int) * (Nb + 1)));
+    HIP_CHECK(hipMalloc((void**)&d_Acols, sizeof(rocsparse_int) * nnzb));
+    HIP_CHECK(hipMalloc((void**)&d_Avals, sizeof(double) * nnz));
+    HIP_CHECK(hipMalloc((void**)&d_x, sizeof(double) * N));
+    HIP_CHECK(hipMalloc((void**)&d_b, sizeof(double) * N));
+
+    if (useJacMatrix) {
+        HIP_CHECK(hipMalloc((void**)&d_Mrows, sizeof(rocsparse_int) * (Nb + 1)));
+        HIP_CHECK(hipMalloc((void**)&d_Mcols, sizeof(rocsparse_int) * nnzbs_prec));
+        HIP_CHECK(hipMalloc((void**)&d_Mvals, sizeof(double) * nnzbs_prec * block_size * block_size));
+    } else { // preconditioner matrix is same
+        HIP_CHECK(hipMalloc((void**)&d_Mvals, sizeof(double) * nnzbs_prec * block_size * block_size));
+        d_Mcols = d_Acols;
+        d_Mrows = d_Arows;
+    }
+
+    initialized = true;
+} // end initialize()
+
+template <unsigned int block_size>
+void rocsparseSolverBackend<block_size>::copy_system_to_gpu(double *b) {
+    Timer t;
+
+#if COPY_ROW_BY_ROW
```

Comment:
What is the default?

Reply:
Fixed in a squash.
The `BdaBridge` has `checkMemoryContiguous()`, this will error before copying row-by-row could be done.